core/src/main/java/io/grpc/transport/MessageFramer.java

```diff
@@ -270,26 +317,53 @@ public void write(byte[] b, int off, int len) {
   }
 
   /**
-   * Implements the same general contract of DeferredInputStream, although is unable to extend it.
+   * Produce a collection of {@link WritableBuffer} instances from the data written to an
+   * {@link OutputStream}.
    */
-  private static class DeferredByteArrayInputStream extends ByteArrayInputStream {
-    public DeferredByteArrayInputStream(byte[] buf, int offset, int length) {
-      super(buf, offset, length);
+  private class BufferChainOutputStream extends OutputStream {
+
+    private final byte[] singleByte = new byte[1];
+    private List<WritableBuffer> bufferList;
+    private WritableBuffer current;
+
+    private BufferChainOutputStream() {
+      bufferList = new ArrayList<WritableBuffer>();
```

Comment:
Would there be any advantage to lazy creating the `bufferList`? I'm guessing not.